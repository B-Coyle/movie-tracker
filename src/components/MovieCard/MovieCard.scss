.movie-card {
  background-position: center;
  background-repeat: no-repeat;
  background-size: cover;
  position: relative;
  overflow: hidden;
  height: 400px;
  width: 250px;
  margin: 5px;
  transition: 0.3s;
	width: 280px;
	height: 360px;
	border-radius: 15px;
	padding: 1.5rem;
	position: relative;
  display: flex;
  cursor: pointer;
	align-items: flex-end;
	transition: 0.4s ease-out;
	box-shadow: 0px 7px 10px rgba(black, 0.5);
	&:hover {
		transform: translateY(-20px);
  }	&:before {
			opacity: 1;
  }	
}