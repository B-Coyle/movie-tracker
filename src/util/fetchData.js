export const fetchData = async(url, options) => {
    const response = await fetch(url, options)
    if(response.ok) {
        return response.json()
    } else {
        throw Error('Error receiving data from fetch')
    }
}   

export const addUser = async (user) => {
    try  {
        const response = await fetch('http://localhost:3000/api/users/new', {
            method: 'POST',
            body: JSON.stringify(user),
            headers: {
            'Content-Type': 'application/json'
            }
        })
        const data = await response.json()
        return data
    } catch(error) {
        throw new Error()
    }
}
   