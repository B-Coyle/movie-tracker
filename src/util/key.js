export const key = "fb457ec4329812194b77f8e9822b2001";
