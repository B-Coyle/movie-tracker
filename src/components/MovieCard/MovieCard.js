import React, { Component } from 'react'
import { PropTypes } from 'prop-types'
import { connect } from 'react-redux'
import './MovieCard.scss';
import * as actions from '../../util/fetchData';

export class MovieCard extends Component {
  constructor(){
    super()
    this.state = {
      favorite: false
    }
  }

   componentDidMount = async () => { 
    console.log(this.props.movie)
    const currentFavorites = await actions.getFavorites(this.props.user_id)
    const locatedFavorites = currentFavorites.data.find((favoriteMovie) => {
      
      return favoriteMovie.movie_id === this.props.movie.movie_id || favoriteMovie.movie_id === this.props.movie.id
    })
    if (locatedFavorites) {
      this.setState({favorite: true})
    } else {
      this.setState({favorite: false})
    }
  }

  toggleMovieFavorite = async () => {
    let { movie, user_id } = this.props
    let movieID = movie.id
    console.log(movie.id)
 
    if (this.state.favorite) {
      movieID = movie.movie_id
    }
    
    if(this.state.favorite) {
      await actions.deleteFavorite(user_id, movie.id)
    } else {
      await actions.addFavorite(movie.id, user_id, movie)
    }
    
    this.setState({favorite: !this.state.favorite})
  }




  render() {
      return(
          <article className='movie-card' style={{ backgroundImage: `url(${this.props.movie.poster_path})` }}>
              <i onClick={this.toggleMovieFavorite} className={`fas fa-star ${this.state.favorite && 'favorite'}`}></i>
              <div className='info'>
                <p>{this.props.movie.title}</p>
                <p>Rating: {this.props.movie.vote_average}</p>
                <p>Release: {this.props.movie.release_date}</p>
                <p>Summary: {this.props.movie.overview}</p>
              </div>
          </article>
      )
  }
}



export const mapStateToProps = (state) =>({
  user_id: state.users.id
 
})


export default connect(mapStateToProps)(MovieCard)