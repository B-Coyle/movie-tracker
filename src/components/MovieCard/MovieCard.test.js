import React from 'react';
import ReactDOM from 'react-dom';
import MovieCard from './MovieCard';
import {shallow } from 'enzyme';
import {mapStateToProps} from './MovieCard'
import * as actions from '../../util/fetchData';


jest.mock('../../util/fetchData')


describe('MovieCard', () => {
    let wrapper;
    let mockMovie;
    
    beforeEach(() => {
      mockMovie={
        title: 'Avengers',
        rating: 10,
        id: 1,
        releaseDate: 2019,
        synopsis: 'a darn good movie',
        posterImage:  undefined 

      }
      wrapper = shallow(
        <MovieCard movie={mockMovie} user_id={1}/>
      );
  
    })

    describe('MovieCard Component', () => {

      it('should match the snapshot with all data passed in correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });

      describe('ComponentDidMount', () => {
        it('should call getFavorites on componentDidMount', async () => {
          actions.getFavorites.mockImplementation(() => ({data: [mockMovie]}))
          wrapper = shallow(<MovieCard movie={mockMovie} user_id={1}/>, { disableLifecycleMethods: true })
          await wrapper.instance().componentDidMount()
          expect(actions.getFavorites).toHaveBeenCalled()
        });


        it('set favorite to true if its a current favorite', async () => {
          actions.getFavorites.mockImplementation(() => ({data: [{movie_id: 1}]}))
          wrapper = shallow(<MovieCard movie={mockMovie} user_id={1}/>, { disableLifecycleMethods: true })
          await wrapper.instance().componentDidMount()
          expect(wrapper.state().favorite).toBe(true)
        });

        it('set favorite to false if its a current favorite', async () => {
          actions.getFavorites.mockImplementation(() => ({data: [{movie_id: 2}]}))
          wrapper = shallow(<MovieCard movie={mockMovie} user_id={1}/>, { disableLifecycleMethods: true })
          await wrapper.instance().componentDidMount()
          expect(wrapper.state().favorite).toBe(false)
        })

      })
    })
})
