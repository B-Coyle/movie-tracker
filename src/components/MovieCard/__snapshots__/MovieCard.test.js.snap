// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard MovieCard should match the snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <ConnectFunction
    movie={
      Object {
        "id": 1,
        "overview": "a darn good movie",
        "poster_path": undefined,
        "release_date": 2019,
        "title": "Avengers",
        "vote_average": 10,
      }
    }
    user_id={1}
  />
</ContextProvider>
`;
